### ISSUE TITLE
Concatenating tuples does not simplify

### ISSUE THREAD

**Issue by NeilGirdhar** [astral-sh/ty#3646]
### Summary

```py
from typing import assert_type

type Path = tuple[str, ...]


def f() -> tuple[Path, str]:
    return ("a", "b"), "c"


x, y = f()
z = (*x, y)
assert_type(z, Path)
```

### Version

0.0.42

**Comment by carljm** [comment]
It looks like only mypy does this simplification, no other type checker does?

The simplification loses precision. `tuple[*tuple[str, ...], str]` is not equivalent to `tuple[str, ...]`, because the empty tuple inhabits the latter type but not the former (the former type must be at least length 1). And ty understands this distinction; note the revealed `Never` here: https://play.ty.dev/846f247f-2dac-4b92-98ae-ec0fa772b0db

Given that ty also understands that `tuple[*tuple[str, ...], str]` is a subtype of `tuple[str, ...]`, is keeping the more precise type causing you any problem, other than not seeing the type you expect in an `assert_type`? (This is the problem with `assert_type` -- it's often too strict an assertion, since it requires type equivalence.)

Open to arguments, but it doesn't seem worth it to me to make this lossy simplification.

**Comment by NeilGirdhar** [comment]
I agree with everything you're saying, and I'm pretty sure I just got the covariance/contravariance backwards somewhere:
```
error[invalid-argument-type]: Argument to bound method `Assembler.assemble` is incorrect
   --> efax/_src/structure/assembler.py:176:30
    |
176 |         return self.assemble(d)
    |                              ^ Expected `Mapping[Path, JaxComplexArray]`, found `dict[tuple[*tuple[str, ...], str], JaxComplexArray]`
    |
```
Thanks for explaining.

**Comment by carljm** [comment]
Oh, I see -- what you're running into there is the invariance of `dict` in its key type, which is indeed a case where having a more-precise-but-not-equivalent type can cause problems. Hard to say without seeing the full code, but usually you can add an annotation somewhere that will convince us (via bidirectional inference) to fall back to the less precise supertype.

**Comment by NeilGirdhar** [comment]
Yes, thanks, that's exactly what I had to do. 😄
